src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java

```diff
@@ -29,168 +29,415 @@
+            // we have a match
+            setServerUrl(url);
             return;
         }
-        this.bitbucketServerUrl = Util.fixEmpty(url);
-        if (this.bitbucketServerUrl != null) {
-            // Remove a possible trailing slash
-            this.bitbucketServerUrl = this.bitbucketServerUrl.replaceAll("/$", "");
-        }
-        resetId();
+        LOGGER.log(Level.WARNING, "Call to legacy setBitbucketServerUrl({0}) method is configuring an url missing "
+                + "from the global configuration.", url);
+        setServerUrl(url);
     }
 
+    @Deprecated
+    @Restricted(DoNotUse.class)
+    @RestrictedSince("2.2.0")
     @CheckForNull
     public String getBitbucketServerUrl() {
-        return bitbucketServerUrl;
+        if (BitbucketEndpointConfiguration.get().findEndpoint(serverUrl) instanceof BitbucketCloudEndpoint) {
+            return null;
+        }
+        return serverUrl;
     }
 
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
+    @NonNull
     public String getIncludes() {
-        return includes;
+        for (SCMTrait<?> trait : traits) {
+            if (trait instanceof WildcardSCMHeadFilterTrait) {
+                return ((WildcardSCMHeadFilterTrait) trait).getIncludes();
+            }
+        }
+        return "*";
     }
 
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
     @DataBoundSetter
-    public void setIncludes(String includes) {
-        this.includes = includes;
+    public void setIncludes(@NonNull String includes) {
+        for (int i = 0; i < traits.size(); i++) {
+            SCMTrait<?> trait = traits.get(i);
+            if (trait instanceof WildcardSCMHeadFilterTrait) {
+                WildcardSCMHeadFilterTrait existing = (WildcardSCMHeadFilterTrait) trait;
+                if ("*".equals(includes) && "".equals(existing.getExcludes())) {
+                    traits.remove(i);
+                } else {
+                    traits.set(i, new WildcardSCMHeadFilterTrait(includes, existing.getExcludes()));
+                }
+                return;
+            }
+        }
+        if (!"*".equals(includes)) {
+            traits.add(new WildcardSCMHeadFilterTrait(includes, ""));
+        }
     }
 
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
+    @NonNull
     public String getExcludes() {
-        return excludes;
+        for (SCMTrait<?> trait : traits) {
+            if (trait instanceof WildcardSCMHeadFilterTrait) {
+                return ((WildcardSCMHeadFilterTrait) trait).getExcludes();
+            }
+        }
+        return "";
     }
 
+    @Deprecated
+    @Restricted(NoExternalUse.class)
+    @RestrictedSince("2.2.0")
     @DataBoundSetter
-    public void setExcludes(String excludes) {
-        this.excludes = excludes;
+    public void setExcludes(@NonNull String excludes) {
+        for (int i = 0; i < traits.size(); i++) {
```

Comment:
I'll think about that

- [ ] @stephenc to decide whether to add a monitor lock